Initial draft for C# (.NET) quickstart (#6092)
github/docs: content/codespaces/getting-started-with-codespaces/getting-started-with-your-dotnet-project.md

---

> title: Getting started with your C# (.NET) project in Codespaces
> shortTitle: Getting started with your C# (.NET) project

Reviewer: ```suggestion
title: Getting started with your C# (.NET) project in Codespaces
shortTitle: Getting started with your C# (.NET) project
allowTitleToDifferFromFilename: true
```

intro: 'You can create a custom dev container with all the tools necessary to get started with your C# (.NET) project in {% data variables.product.prodname_codespaces %}.'
product: '{% data reusables.gated-features.codespaces %}'
versions:
  free-pro-team: '*'
topics:
  - Codespaces
---

### Introduction

This guide shows you how to set up your C# (.NET) project in {% data variables.product.prodname_codespaces %}. It will take you through an example of opening your project in a codespace, and adding and modifying a dev container configuration from a template.

#### Prerequisites 

> - You should have an existing C# (.NET) project in a repository on {% data variables.product.prodname_dotcom_the_website %}. If you don't have a project, you can try this tutorial with the following example: https://github.com/2percentsilk/dotnet-quickstart.

Author: Open to other sample repos - we want it to be relatively simple so I just used `dotnet new` to generate this application.

- You must have {% data variables.product.prodname_codespaces %} enabled for your organization.

### Step 1: Open your project in a codespace

1. Navigate to your project's repository. Use the {% octicon "download" aria-label="The download icon" %} **Code** drop-down menu, and select **Open with Codespaces**. If you don’t see this option, your project isn’t available for {% data variables.product.prodname_codespaces %}.
  
  ![Open with Codespaces button](/assets/images/help/codespaces/open-with-codespaces-button.png) 

2. To create a new codespace, click {% octicon "plus" aria-label="The plus icon" %} **New codespace**.
  ![New codespace button](/assets/images/help/codespaces/new-codespace-button.png)

> When you create a codespace, your project is created on a remote VM that is dedicated to you. By default, the container for your codespace has many languages and runtimes including Node.js, JavaScript, Typescript, nvm, npm, and yarn. It also includes a common set of tools like git, wget, rsync, openssh, and nano. 

Reviewer: ```suggestion
When you create a codespace, your project is created on a remote VM that is dedicated to you. By default, the container for your codespace has many languages and runtimes including .NET. It also includes a common set of tools like git, wget, rsync, openssh, and nano. 
```

> You can customize your codespace by adjusting the amount of vCPUs and RAM], [adding dotfiles to personalize your environment](/codespaces/setting-up-your-codespace/personalizing-codespaces-for-your-account), or by modifying the tools and scripts installed. 

Reviewer: ```suggestion
You can customize your codespace by adjusting the amount of vCPUs and RAM, [adding dotfiles to personalize your environment](/codespaces/setting-up-your-codespace/personalizing-codespaces-for-your-account), or by modifying the tools and scripts installed. 
```

{% data variables.product.prodname_codespaces %} uses a file called `devcontainer.json` to store configurations. On launch {% data variables.product.prodname_codespaces %} uses the file to install any tools, dependencies, or other set up that might be needed for the project. For more information, see "[Configuring Codespaces for your project](/codespaces/setting-up-your-codespace/configuring-codespaces-for-your-project)."

> The next section shows you how to modify your tools by adding a dev container.

Reviewer: ```suggestion
```

### Step 2: Add a dev container to your codespace from a template  

The default codespaces container comes with the latest .NET version and common tools preinstalled. However, we encourage you to set up a custom container so you can tailor the tools and scripts that run as part of codespace creation to your project's needs and ensure a fully reproducible environment for all {% data variables.product.prodname_codespaces %} users in your repository.

> To set up your project with a custom container, you will need to use a `devcontainer.json` file to define the environment. In {% data variables.product.prodname_codespaces %} you can add this either from a template or you can create your own. For more information on dev containers, see [Configuring your codespace](/codespaces/setting-up-your-codespace/configuring-codespaces-for-your-project). 

Reviewer: ```suggestion
To set up your project with a custom container, you will need to use a `devcontainer.json` file to define the environment. In {% data variables.product.prodname_codespaces %} you can add this either from a template or you can create your own. For more information on dev containers, see "[Configuring Codespaces for your project
](/codespaces/setting-up-your-codespace/configuring-codespaces-for-your-project)."
```

> This example guides you through adding a `devcontainer.json` file from a template.

Reviewer: ```suggestion
```

	// 
	// 2. Uncomment these 'remoteEnv' lines:
	//    "remoteEnv": {
	// 	      "ASPNETCORE_Kestrel__Certificates__Default__Password": "SecurePwdGoesHere",
	//        "ASPNETCORE_Kestrel__Certificates__Default__Path": "/home/vscode/.aspnet/https/aspnetapp.pfx",
	//    },
	//
	// 3. Do one of the following depending on your scenario:
	//    * When using GitHub Codespaces and/or Remote - Containers:
	//      1. Start the container
	//      2. Drag ~/.aspnet/https/aspnetapp.pfx into the root of the file explorer
	//      3. Open a terminal in VS Code and run "mkdir -p /home/vscode/.aspnet/https && mv aspnetapp.pfx /home/vscode/.aspnet/https"
	//
	//    * If only using Remote - Containers with a local container, uncomment this line instead:
	//      "mounts": [ "source=${env:HOME}${env:USERPROFILE}/.aspnet/https,target=/home/vscode/.aspnet/https,type=bind" ],

	// Use 'postCreateCommand' to run commands after the container is created.
	// "postCreateCommand": "dotnet restore",

	// Comment out connect as root instead. More info: https://aka.ms/vscode-remote/containers/non-root.
	"remoteUser": "vscode"
}
```

> - **Name** - We can name our dev container anything, this is just the default
> - **Build** - Our build properties
>   - **Dockerfile** - In our build object, Dockerfile is a reference to the Dockerfile in the same folder that was the second file added to our project. This is the reference path.
>   - **Args**
>     - **Variant**: We only have one build argument here which is the node variant we want to use which is passed into our Dockerfile.
> - **Settings** - These are {% data variables.product.prodname_vscode %} settings we wish to set
>   - **Terminal.integrated.shell.linux** - While bash is the default here, we could use zsh for example by modifying this.
> - **Extensions** - These are extensions included by default.
>   - **ms-python.python** - The Microsoft Python extension provides rich support for the Python language (for all actively supported versions of the language: >=3.6), including features such as IntelliSense, linting, debugging, code navigation, code formatting, refactoring, variable explorer, test explorer, and more.
> - **forwardPorts** - By default we can forward a port, like port 3000, but these will also forward automatically
> - **postCreateCommand** - If we want to run anything after we land in our codespace that’s not defined in our Dockerfile, like `dotnet restore`, we can do that here.
> - **remoteUser** - We’re running as the vscode user, but you can optionally set this to root

Reviewer: ```suggestion
- **Name** - You can name our dev container anything, this is just the default.
- **Build** - The build properties.
  - **Dockerfile** - In the build object, `dockerfile` is a reference to the Dockerfile that was also added from the template.
  - **Args**
    - **Variant**: This file only contains one build argument, which is the .NET Core version that we want to use.
- **Settings** - These are {% data variables.product.prodname_vscode %} settings.
  - **Terminal.integrated.shell.linux** - While bash is the default here, you could use other terminal shells by modifying this.
- **Extensions** - These are extensions included by default.
  - **ms-dotnettools.csharp** - The Microsoft C# extension provides rich support for developing in C#, including features such as IntelliSense, linting, debugging, code navigation, code formatting, refactoring, variable explorer, test explorer, and more.
- **forwardPorts** - Any ports listed here will be forwarded automatically.
- **postCreateCommand** - If you want to run anything after you land in your codespace that’s not defined in the Dockerfile, like `dotnet restore`, you can do that here.
- **remoteUser** - By default, you’re running as the vscode user, but you can optionally set this to root.
```

#     && apt-get -y install --no-install-recommends <your-package-list-here>

# [Optional] Uncomment this line to install global node packages.
# RUN su vscode -c "source /usr/local/share/nvm/nvm.sh && npm install -g <your-package-here>" 2>&1
```

You can use the Dockerfile to add additional container layers to specify OS packages, node versions, or global packages we want included in our container.

### Step 3: Modify your devcontainer.json file 

With your dev container added and a basic understanding of what everything does, you can now make changes to configure it for your environment. In this example, you'll add properties to install extensions and restore your project dependancies when your codespace launches.

1. In the Explorer, expand the `.devcontainer` folder and select the `devcontainer.json` file from the tree to open it. 

  !["Codespaces: Rebuild Container" in the command palette](/assets/images/help/codespaces/devcontainers-options.png)  

2. Update your the `extensions` list in your `devcontainer.json` file to add a few extensions that are useful when working with your project.
   
  ```json{:copy} 
  "extensions": [

> 		  "ms-dotnettools.csharp",

Author: Any extensions you might suggest that are common/helpful?

		  "streetsidesoftware.code-spell-checker",
	  ],
  ```

3. Uncomment the `postCreateCommand` to restore dependencies as part of the codespace setup process.

  ```json{:copy} 
  // Use 'postCreateCommand' to run commands after the container is created.
  "postCreateCommand": "dotnet restore",
  ```

4. To rebuild your container and apply the devcontainer.json changes, access the command palette (`shift command P` / `shift control P`), then start typing "rebuild". Click **Codespaces: Rebuild Container**. 

  ![Rebuild container option](/assets/images/help/codespaces/codespaces-rebuild.png)

  Rebuilding inside your codespace ensures your changes work as expected before you commit the changes to the repository. If something does result in a failure, you’ll be placed in a codespace with a recovery container that you can rebuild from to keep adjusting your container.

> 5. Check your changes were successfully applied by verifying the Code Spell Checker extension was installed.

Reviewer: ```suggestion
5. Check your changes were successfully applied by verifying the "Code Spell Checker" extension was installed.
```

> ![Extensions list](/assets/images/help/codespaces/dotnet-extensions.png) 

Reviewer: ```suggestion
    ![Extensions list](/assets/images/help/codespaces/dotnet-extensions.png) 
```

   
### Step 4: Run your application

In the previous section, you used the `postCreateCommand` to installing a set of packages via pip3. With our dependencies now installed, we can run our application.

1. Run your application by pressing `F5` or entering `dotnet watch run` in your terminal.

2. When your project starts, you should see a toast in the bottom right corner with a prompt to connect to the port your project uses. 

  ![Port forwarding toast](/assets/images/help/codespaces/python-port-forwarding.png)

### Step 5: Commit your changes

> Once you've made changes to your codespace, either new code or configuration changes, you'll want to commit your changes. Committing changes to your repository ensures that anyone else who creates a codespace from this repository has the same configuration. This also means that any customization you do, such as adding {% data variables.product.prodname_vscode %} extensions, will appear for all users.
> 
> 1. In the Activity Bar, click on the **Source Control** view.
> 
>   ![Source control view](/assets/images/help/codespaces/codespaces-commit-activity.png)  
> 
> 2. To stage your changes, click  **+**.
> 
>   ![Explorer with staging button highlighted](/assets/images/help/codespaces/codespaces-commit-stage.png)
> 
> 3. Type a commit message and then use `Ctrl+Enter` / `cmd+Enter` to commit the changes.  
> 
>   ![Explorer with commit message added](/assets/images/help/codespaces/codespaces-commit-commit-message.png)  
> 
> 4. To create a PR, click the PR icon. 
> 
>   ![Explorer with staging button highlighted](/assets/images/help/codespaces/codespaces-commit-pr-button.png)  
> 
> 5. Select the branches that you want to merge into, then click **Create**.
> 
>   ![Explorer with staging button highlighted](/assets/images/help/codespaces/codespaces-commit-pr.png)  

Reviewer: ```suggestion
{% data reusables.codespaces.committing-link-to-procedure %} 
```
